api/reporting/schemas/reporting-codegen.rb

```diff
@@ -136,68 +261,190 @@ def to_swagger_property(column)
+    + "  List<#{TABLE_INFO[:projection_interface]}> getReporting#{to_camel_case(TABLE_INFO[:name], true)}s();"
+end
+
+sql = to_query
+
+write_output(OUTPUTS[:projection_query], sql, 'Projection Query')
+
+# Unit Test Constants
+#
+BASE_TIMESTAMP = Time.new(2015, 5, 5).freeze
+TIMESTAMP_DELTA_SECONDS = 24 * 60 * 60 # .freeze # seconds in day
+
+# N.B. some Short fields are only valid up to the number of associated enum values - 1. Fixing these
+# up by hand for now.
+def to_constant_declaration(column, index)
+  value = case column[:java_type]
+          when 'String'
+            "\"foo_#{index}\""
+          when 'Integer'
+            "%d" % [index]
+          when 'Long'
+            "%dL" % [index]
+          when 'Double'
+            "%f" % [index + 0.5]
+          when 'Boolean'
+            index.even? # just flip it every time
+          when 'Timestamp'
+            # add a day times the index to base timestamp
+            timestamp = BASE_TIMESTAMP + TIMESTAMP_DELTA_SECONDS * index
+            "Timestamp.from(Instant.parse(\"#{timestamp.strftime("%Y-%m-%dT00:00:00.00Z")}\"))"
+          else
+            if column[:is_enum]
+              column[:default_enum]
+            else
+              index.to_s
+            end
+          end
+  "public static final #{column[:java_type]} #{column[:java_constant_name]} = #{value};"
+end
+
+constants = COLUMNS.enum_for(:each_with_index) \
+  .map { |col, index| to_constant_declaration(col, index) } \
+  .join("\n")
+
+write_output(OUTPUTS[:unit_test_constants], constants, 'Unit Test Constants')
+
+### Mock Instantiation
+# Mock the projection interface for testing with mock services exposing them
+mocks = COLUMNS.map { |col|
+  "doReturn(#{col[:java_constant_name]}).when(#{TABLE_INFO[:mock]}).#{col[:getter]}();"
+}
+
+lines = []
+lines << "final #{TABLE_INFO[:projection_interface]} #{TABLE_INFO[:mock]} = mock(#{TABLE_INFO[:projection_interface]}.class);"
+lines << mocks
+lines.flatten!
+
+write_output(OUTPUTS[:unit_test_mocks], lines.join("\n"), 'Unit Test Mocks')
+
+### Assertions
+dto_assertions = COLUMNS.map{ |col|
+  getter_call = "#{to_camel_case(table_name, false)}.#{col[:getter]}()"
+  expected = col[:java_constant_name]
+  if col[:java_type].eql?('Timestamp')
+    "    assertTimeApprox(#{getter_call}, #{expected});"
+  else
+    "    assertThat(#{getter_call}).isEqualTo(#{expected});"
+  end
+
+}.join("\n")
+
+write_output(OUTPUTS[:dto_assertions], dto_assertions, 'Unit Test DTO Assertions')
+
+### Parameter Column Enum
+def object_value_function(col)
+  if col[:is_enum]
+    "#{col[:lambda_var]} -> enumToString(#{col[:lambda_var]}.#{col[:getter]}())"
+  else
+    case col[:java_type]
+    when 'Timestamp' # actually OffsetDateTime on the DTO
+      "#{col[:lambda_var]} -> toInsertRowString(#{col[:lambda_var]}.#{col[:getter]}())"
+    else
+      "#{TABLE_INFO[:dto_class]}::#{col[:getter]}"
+    end
+  end
+end
+
+# enum types are special
+JAVA_TYPE_TO_QPV_FACTORY = {
```

Comment:
Each column knows the function needed to extract either a QueryParameterValue (not null), or an insertable Object (or null so we don't add it to the map). These errors only show up when dealing with BigQuery itself.

TODO: look into writing a BigQueryTest for this system.